byway/src/main/webapp/routepage.js

```diff
@@ -11,171 +11,206 @@
   let request = {
     origin:  start,
     destination: end,
-    travelMode: 'DRIVING'
+    travelMode: 'DRIVING',
+    waypoints: Array.from(waypoints),
+    optimizeWaypoints: true
   };
   directionsService.route(request, function(response, status) {
     if (status == 'OK') {
       directionsRenderer.setDirections(response);
+      computeTotalDistance(response);
     } else {
       window.alert("Could not calculate route due to: " + status);
     }
   });
 }
 
+/**
+ * Calculates and sums up the distance and time duration between all destinations (legs)
+ * @param {response} result response from the directions service object
+ */
+function computeTotalDistance(result) {
+  var totalDist = 0;
+  var totalTime = 0;
+  // full route
+  var route = result.routes[0];
+  for (i = 0; i < route.legs.length; i++) {
+    totalDist += route.legs[i].distance.value;
+    totalTime += route.legs[i].duration.value;
+  }
+
+  totalDist = (totalDist / 1000).toFixed(2);
+  hours = Math.floor(totalTime / 3600);
+  minutes = Math.round((totalTime - hours*3600) / 60);
+ 
+  document.getElementById("distance").innerText = totalDist + "km";
+  document.getElementById("duration").innerText = 
+    (hours == 0) ? minutes + " mins" : hours + " hr " + minutes + " mins";
+}
+ 
 /** Clear the stops panel in the html */
 function clearStops(){
   const stopList = document.getElementById('stop-list');
   if(stopList != null){
     stopList.innerText = "";
   }
 }
-
+ 
 /** Get the new list of stops from datastore onload */
 function getStopsOnload(){
   clearStops();
   fetch('/api/stop')
   .then(response => response.json())
   .then((stopsResponse) => {
-    stopsResponse.forEach((stop)=>{
-      stops.push(stop);
-    });
+    if(stopsResponse != null){
+      stopsResponse.forEach((stop)=>{
+        stops.add(stop);
+        waypoints.add({location:stop});
+      });
+    }
+    calcRoute();
     renderStopsList();
   });
 }
-
+ 
 /** Re-render stop list synchronously */
 function renderStopsList(){
   clearStops();
+  const stopList = document.getElementById('stop-list');
   stops.forEach((stop)=>{
-    renderStop(stop);
+    stopList.appendChild(createStopButton(stop));
   })
 }
-
+ 
 /** Render stop list  
  *  @param {String} stop a String to add as a button in the schedule panel in the html
+ *  @return {button} stopBtn a button showing a selected stop
  */
-function renderStop(stop){
-  const stopList = document.getElementById('stop-list');
-  let btn = document.createElement('button');
-  btn.innerText = stop;
-  btn.className =  "btn rec-btn";
-  btn.addEventListener("click", function() {
-    updateStops(stop, "remove");
+function createStopButton(stop){
+  const stopBtn = document.createElement('button');
+  stopBtn.innerText = stop;
+  stopBtn.className =  "btn rec-btn";
+  stopBtn.addEventListener("click", function() {
+    // remove place from stops and waypoints set
+    stops = new Set([...stops].filter(stopObj => stopObj != stop));
```

Comment:
For the stops set, you could just delete the single stop instead instead of filtering the whole array. 
```javascript
stops.delete(stop);
```

Reply:
Oh yeah you're absolutely right that's a funny error, changed.